File: src/733. Flood Fill.cpp
class Solution {
private:
    void dfs(vector<vector<int>>& image, int r, int c, int oldColor, int newColor) {
        if (
            r < 0 ||
            r >= image.size() ||
            c < 0 ||
            c >= image[0].size() ||
            image[r][c] == newColor ||
            image[r][c] != oldColor
        ) return;
        image[r][c] = newColor;
        dfs(image, r - 1, c, oldColor, newColor);
        dfs(image, r + 1, c, oldColor, newColor);
        dfs(image, r, c - 1, oldColor, newColor);
        dfs(image, r, c + 1, oldColor, newColor);
    }
public:
    vector<vector<int>> floodFill(vector<vector<int>>& image, int sr, int sc, int newColor) {
        dfs(image, sr, sc, image[sr][sc], newColor);
        return image;
    }
};

File: src/733. Flood Fill.py
class Solution:
    def floodFill(self, image: List[List[int]], sr: int, sc: int, color: int) -> List[List[int]]:
        R = len(image)
        C = len(image[0])
        old_color = image[sr][sc]
        def dfs(r: int, c: int):
            if r < 0 or r >= R or c < 0 or c >= C or image[r][c] == color or image[r][c] != old_color:
                return
            image[r][c] = color
            dfs(r - 1, c)
            dfs(r + 1, c)
            dfs(r, c - 1)
            dfs(r, c + 1)
        dfs(sr, sc)
        return image